open-craft/opencraft
# [BB-4408] Allow creating instances with domain like "subdomain.sandbox.opencraft.hosting"

This PR addresses [BB-4408](https://tasks.opencraft.com/browse/BB-4408).

**Issues addressed in this PR:**
Provisioning sandbox servers with domain names of the form `subdomain.sandbox.opencraft.hosting` is a common use case for developers and in automated sandbox builds. However, PR #737 introduced validation that renders this impossible. 

**Affected components:**
This PR affects the instance factory methods `instance_factory` and `production_instance_factory`.

**Test instructions:**
Create `OpenEdXInstance`s as describes [here](https://ocim.opencraft.com/en/latest/provisioning-sandboxes/#manual-provisioning).

1. Create an instance using `instance_factory` with an arbitrary subdomain. This should succeed.
2. Create an instance using `production_instance_factory` with uppercase letters in the subdomain. This should succeed, but the subdomain should be coerced to lowercase.
3. Create an instance using `production_instance_factory` that has the same subdomain as in step 2, but all lowercase. This should fail due to another instance already having the same subdomain.
